src/backups.py

```diff
@@ -0,0 +1,307 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of backups."""
+
+import datetime
+import logging
+from typing import Dict, List, Tuple
+
+from charms.data_platform_libs.v0.s3 import S3Requirer
+from charms.mysql.v0.mysql import (
+    MySQLGetMemberStateError,
+    MySQLSetInstanceOfflineModeError,
+    MySQLSetInstanceOptionError,
+)
+from ops.charm import ActionEvent, CharmBase
+from ops.framework import Object
+from ops.jujuversion import JujuVersion
+
+from mysql_k8s_helpers import MySQLExecuteBackupCommandsError
+from s3_helpers import list_backups_in_s3_path, upload_content_to_s3
+
+logger = logging.getLogger(__name__)
+
+MYSQL_BACKUPS = "mysql-backups"
+
+
+class MySQLBackups(Object):
+    """Encapsulation of backups for MySQL."""
+
+    def __init__(self, charm: CharmBase, s3_integrator: S3Requirer) -> None:
+        super().__init__(charm, MYSQL_BACKUPS)
+
+        self.charm = charm
+        self.s3_integrator = s3_integrator
+
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups)
+
+    def _on_list_backups(self, event: ActionEvent) -> None:
+        """Handle the list backups action.
+
+        List backups available to restore by this application.
+        """
+        try:
+            logger.info("Retrieving s3 parameters from the s3-integrator relation")
+            s3_parameters, missing_parameters = self._retrieve_s3_parameters()
+            if missing_parameters:
+                event.fail(f"Missing S3 parameters: {missing_parameters}")
+                return
+
+            logger.info("Listing backups in the specified s3 path")
+            backup_ids = list_backups_in_s3_path(
+                s3_parameters["bucket"],
+                s3_parameters["path"],
+                s3_parameters["region"],
+                s3_parameters["endpoint"],
+                s3_parameters["access-key"],
+                s3_parameters["secret-key"],
+            )
+            event.set_results({"backup-ids": backup_ids})
+        except Exception:
+            event.fail("Failed to retrieve backup ids from S3")
```

Comment:
We haven't really discussed our philosophy for using try/except in code. The one thing we have talked about:
When invoking a method in a helper class, the helper class should raise custom errors. This is so the invoker can decide how to handle the exception (making it possible to set unit status, like BlockedStatus, at the charm.py file instead of the helper file)